File: packages/dart/test/ply_export_test.dart
import 'dart:typed_data';
import 'package:openskp/openskp.dart';
import 'package:test/test.dart';

void main() {
  group('PLY Exporter (ASCII & Binary)', () {
    test('serializes Scene to ASCII PLY text format', () {
      final scene = Scene(
        sceneHierarchy: InstanceNode(
          name: 'Root',
          definitionName: 'RootDef',
          layer: 'Layer0',
          positionMm: (0.0, 0.0, 0.0),
          properties: {},
          children: [],
        ),
        meshIndex: {},
        glbPrimitives: [
          GlbPrimitive(
            geomName: 'Box',
            materialIndex: 0,
            positions: Float32List.fromList([0, 0, 0, 1, 0, 0, 0, 1, 0]),
            normals: Float32List.fromList([0, 0, 1, 0, 0, 1, 0, 0, 1]),
            uvs: Float32List.fromList([0, 0, 1, 0, 0, 1]),
            indices: Uint32List.fromList([0, 1, 2]),
          ),
        ],
        gltfMaterials: [],
      );

      final plyText = toPlyAscii(scene);
      expect(plyText, contains('format ascii 1.0'));
      expect(plyText, contains('element vertex 3'));
      expect(plyText, contains('element face 1'));
      expect(plyText, contains('3 0 1 2'));
    });

    test('serializes Scene to Binary PLY byte array', () {
      final scene = Scene(
        sceneHierarchy: InstanceNode(
          name: 'Root',
          definitionName: 'RootDef',
          layer: 'Layer0',
          positionMm: (0.0, 0.0, 0.0),
          properties: {},
          children: [],
        ),
        meshIndex: {},
        glbPrimitives: [
          GlbPrimitive(
            geomName: 'Box',
            materialIndex: 0,
            positions: Float32List.fromList([0, 0, 0, 1, 0, 0, 0, 1, 0]),
            normals: Float32List.fromList([0, 0, 1, 0, 0, 1, 0, 0, 1]),
            uvs: Float32List.fromList([0, 0, 1, 0, 0, 1]),
            indices: Uint32List.fromList([0, 1, 2]),
          ),
        ],
        gltfMaterials: [],
      );

      final data = toPlyBinary(scene);
      final text = String.fromCharCodes(data);
      expect(text, contains('format binary_little_endian 1.0'));
      expect(text, contains('element vertex 3'));
      expect(text, contains('element face 1'));
    });
  });
}

File: packages/cpp/tests/ply_export_test.cpp
#include "openskp/ply_export.hpp"

#include <gtest/gtest.h>

namespace openskp {
namespace {

TEST(PlyExport, SerializesSceneToPlyAsciiText) {
  Scene scene;
  GlbPrimitive prim;
  prim.geom_name = "Box";
  prim.material_index = 0;
  prim.positions = {0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0};
  prim.normals = {0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0};
  prim.uvs = {0.0, 0.0, 1.0, 0.0, 0.0, 1.0};
  prim.indices = {0, 1, 2};
  scene.glb_primitives.push_back(prim);

  std::string ply_text = to_ply_ascii(scene);
  EXPECT_NE(ply_text.find("format ascii 1.0"), std::string::npos);
  EXPECT_NE(ply_text.find("element vertex 3"), std::string::npos);
  EXPECT_NE(ply_text.find("element face 1"), std::string::npos);
  EXPECT_NE(ply_text.find("3 0 1 2"), std::string::npos);
}

TEST(PlyExport, SerializesSceneToPlyBinaryData) {
  Scene scene;
  GlbPrimitive prim;
  prim.geom_name = "Box";
  prim.material_index = 0;
  prim.positions = {0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0};
  prim.normals = {0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0};
  prim.uvs = {0.0, 0.0, 1.0, 0.0, 0.0, 1.0};
  prim.indices = {0, 1, 2};
  scene.glb_primitives.push_back(prim);

  std::vector<std::uint8_t> data = to_ply_binary(scene);
  std::string text(data.begin(), data.end());
  EXPECT_NE(text.find("format binary_little_endian 1.0"), std::string::npos);
  EXPECT_NE(text.find("element vertex 3"), std::string::npos);
  EXPECT_NE(text.find("element face 1"), std::string::npos);
}

}  // namespace
}  // namespace openskp
